File: Services/RecipeService.swift
//
//  RecipeService.swift
//  BellaRecipeRecommender
//
//  Service for fetching recipes from APIs
//

import Foundation

class RecipeService {
    static let shared = RecipeService()
    
    private let networkManager = NetworkManager.shared
    
    // MARK: - TheMealDB Models
    private struct MealSearchResponse: Codable {
        let meals: [MealBasic]?
    }
    
    private struct MealBasic: Codable {
        let idMeal: String
        let strMeal: String
        let strMealThumb: String
    }
    
    private struct MealDetailResponse: Codable {
        let meals: [MealDetail]?
    }
    
    private struct MealDetail: Codable {
        let idMeal: String
        let strMeal: String
        let strCategory: String?
        let strArea: String?
        let strInstructions: String?
        let strMealThumb: String?
        let strSource: String?
        let strTags: String?
        
        // Ingredients (up to 20)
        let strIngredient1: String?
        let strIngredient2: String?
        let strIngredient3: String?
        let strIngredient4: String?
        let strIngredient5: String?
        let strIngredient6: String?
        let strIngredient7: String?
        let strIngredient8: String?
        let strIngredient9: String?
        let strIngredient10: String?
        let strIngredient11: String?
        let strIngredient12: String?
        let strIngredient13: String?
        let strIngredient14: String?
        let strIngredient15: String?
        let strIngredient16: String?
        let strIngredient17: String?
        let strIngredient18: String?
        let strIngredient19: String?
        let strIngredient20: String?
        
        func getIngredients() -> [String] {
            let ingredients = [
                strIngredient1, strIngredient2, strIngredient3, strIngredient4, strIngredient5,
                strIngredient6, strIngredient7, strIngredient8, strIngredient9, strIngredient10,
                strIngredient11, strIngredient12, strIngredient13, strIngredient14, strIngredient15,
                strIngredient16, strIngredient17, strIngredient18, strIngredient19, strIngredient20
            ]
            return ingredients.compactMap { $0?.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty }
        }
    }
    
    // MARK: - Public Methods
    
    func searchRecipesByIngredient(_ ingredient: String) async throws -> [Recipe] {
        let urlString = APIConstants.TheMealDB.searchByIngredient(ingredient)
        guard let url = URL(string: urlString) else {
            throw NetworkManager.NetworkError.invalidURL
        }
        
        let response: MealSearchResponse = try await networkManager.fetch(url: url)
        guard let meals = response.meals else {
            return []
        }
        
        var recipes: [Recipe] = []
        
        for meal in meals {
            if let recipe = try await getRecipeDetails(mealID: meal.idMeal) {
                recipes.append(recipe)
            }
        }
        
        return recipes
    }
    
    func getRecipeDetails(mealID: String) async throws -> Recipe? {
        let urlString = APIConstants.TheMealDB.getRecipeDetails(mealID)
        guard let url = URL(string: urlString) else {
            throw NetworkManager.NetworkError.invalidURL
        }
        
        let response: MealDetailResponse = try await networkManager.fetch(url: url)
        guard let meals = response.meals, let meal = meals.first else {
            return nil
        }
        
        let ingredients = meal.getIngredients()
        
        return Recipe(
            id: meal.idMeal,
            name: meal.strMeal,
            image: meal.strMealThumb ?? "",
            source: meal.strArea ?? "Unknown",
            url: meal.strSource ?? "https://www.themealdb.com",
            ingredients: ingredients,
            instructions: meal.strInstructions,
            prepTime: nil,
            cookTime: nil,
            servings: nil,
            difficulty: determineDifficulty(instructions: meal.strInstructions),
            tags: (meal.strTags?.split(separator: ",").map(String.init)) ?? [],
            nutrition: nil,
            matchedIngredients: 0,
            totalIngredients: ingredients.count
        )
    }
    
    func searchRecipesByMultipleIngredients(_ ingredients: [String]) async throws -> [Recipe] {
        var allRecipes: [Recipe] = []
        var recipeIDs: Set<String> = []
        
        for ingredient in ingredients {
            let recipes = try await searchRecipesByIngredient(ingredient)
            for recipe in recipes {
                if !recipeIDs.contains(recipe.id) {
                    recipeIDs.insert(recipe.id)
                    allRecipes.append(recipe)
                }
            }
        }
        
        return allRecipes
    }
    
    func filterRecipesByDietaryRestrictions(_ recipes: [Recipe], restrictions: [DietaryRestriction]) -> [Recipe] {
        guard !restrictions.isEmpty else { return recipes }
        
        return recipes.filter { recipe in
            restrictions.allSatisfy { restriction in
                matchesDietaryRestriction(recipe: recipe, restriction: restriction)
            }
        }
    }
    
    // MARK: - Private Methods
    
    private func determineDifficulty(instructions: String?) -> Recipe.DifficultyLevel {
        guard let instructions = instructions else { return .medium }
        let wordCount = instructions.split(separator: " ").count
        
        if wordCount < 50 {
            return .easy
        } else if wordCount < 150 {
            return .medium
        } else {
            return .hard
        }
    }
    
    private func matchesDietaryRestriction(recipe: Recipe, restriction: DietaryRestriction) -> Bool {
        let ingredientsLower = recipe.ingredients.map { $0.lowercased() }
        let recipeLower = (recipe.name + recipe.instructions.flatMap { $0 }).lowercased()
        
        switch restriction {
        case .vegan:
            return !ingredientsLower.contains { ingredient in
                ingredient.contains("meat") || ingredient.contains("chicken") ||
                ingredient.contains("fish") || ingredient.contains("egg") ||
                ingredient.contains("milk") || ingredient.contains("cheese") ||
                ingredient.contains("butter") || ingredient.contains("cream")
            }
        case .vegetarian:
            return !ingredientsLower.contains { ingredient in
                ingredient.contains("meat") || ingredient.contains("chicken") ||
                ingredient.contains("fish") || ingredient.contains("beef")
            }
        case .glutenFree:
            return !ingredientsLower.contains { ingredient in
                ingredient.contains("wheat") || ingredient.contains("flour") ||
                ingredient.contains("bread") || ingredient.contains("pasta")
            }
        case .dairyFree:
            return !ingredientsLower.contains { ingredient in
                ingredient.contains("milk") || ingredient.contains("cheese") ||
                ingredient.contains("butter") || ingredient.contains("cream")
            }
        case .keto:
            return recipeLower.contains("keto") || !ingredientsLower.contains { ingredient in
                ingredient.contains("sugar") || ingredient.contains("rice") ||
                ingredient.contains("pasta") || ingredient.contains("bread")
            }
        case .paleo:
            return !ingredientsLower.contains { ingredient in
                ingredient.contains("dairy") || ingredient.contains("grain") ||
                ingredient.contains("legume") || ingredient.contains("processed")
            }
        case .lowCarb:
            return !ingredientsLower.contains { ingredient in
                ingredient.contains("rice") || ingredient.contains("pasta") ||
                ingredient.contains("bread") || ingredient.contains("potato")
            }
        case .highProtein:
            return ingredientsLower.contains { ingredient in
                ingredient.contains("chicken") || ingredient.contains("beef") ||
                ingredient.contains("fish") || ingredient.contains("egg") ||
                ingredient.contains("protein")
            }
        }
    }
}

File: web/src/services/RecipeService.js
import axios from 'axios';

const THEMEALDB_API = 'https://www.themealdb.com/api/json/v1/1';

class RecipeService {
  static async searchRecipesByIngredient(ingredient) {
    try {
      const response = await axios.get(`${THEMEALDB_API}/filter.php?i=${ingredient}`);
      const meals = response.data.meals || [];
      
      // Fetch full details for each meal
      const detailedMeals = await Promise.all(
        meals.map(meal => this.getRecipeDetails(meal.idMeal))
      );
      
      return detailedMeals.filter(meal => meal !== null);
    } catch (error) {
      console.error('Error searching recipes:', error);
      throw error;
    }
  }

  static async getRecipeDetails(mealID) {
    try {
      const response = await axios.get(`${THEMEALDB_API}/lookup.php?i=${mealID}`);
      const meal = response.data.meals?.[0];
      
      if (!meal) return null;

      const ingredients = this.getIngredientsFromMeal(meal);
      
      return {
        id: meal.idMeal,
        name: meal.strMeal,
        image: meal.strMealThumb || '',
        source: meal.strArea || 'Unknown',
        url: meal.strSource || 'https://www.themealdb.com',
        ingredients: ingredients,
        instructions: meal.strInstructions || '',
        prepTime: null,
        cookTime: null,
        servings: null,
        difficulty: this.determineDifficulty(meal.strInstructions),
        tags: meal.strTags ? meal.strTags.split(',') : [],
        nutrition: null,
        matchedIngredients: 0,
        totalIngredients: ingredients.length
      };
    } catch (error) {
      console.error('Error getting recipe details:', error);
      return null;
    }
  }

  static getIngredientsFromMeal(meal) {
    const ingredients = [];
    for (let i = 1; i <= 20; i++) {
      const ingredient = meal[`strIngredient${i}`];
      if (ingredient && ingredient.trim()) {
        ingredients.push(ingredient.trim());
      }
    }
    return ingredients;
  }

  static determineDifficulty(instructions) {
    if (!instructions) return 'Medium';
    const wordCount = instructions.split(/\s+/).length;
    
    if (wordCount < 50) return 'Easy';
    if (wordCount < 150) return 'Medium';
    return 'Hard';
  }

  static async searchRecipesByMultipleIngredients(ingredients) {
    try {
      const allRecipes = [];
      const recipeIDs = new Set();
      
      for (const ingredient of ingredients) {
        try {
          const recipes = await this.searchRecipesByIngredient(ingredient);
          for (const recipe of recipes) {
            if (!recipeIDs.has(recipe.id)) {
              recipeIDs.add(recipe.id);
              allRecipes.push(recipe);
            }
          }
        } catch (error) {
          console.warn(`Error searching for ${ingredient}:`, error);
        }
      }
      
      return allRecipes;
    } catch (error) {
      console.error('Error searching multiple ingredients:', error);
      throw error;
    }
  }

  static filterRecipesByDietaryRestrictions(recipes, restrictions) {
    return recipes.filter(recipe =>
      restrictions.every(restriction =>
        this.matchesDietaryRestriction(recipe, restriction)
      )
    );
  }

  static matchesDietaryRestriction(recipe, restriction) {
    const ingredientsLower = recipe.ingredients.map(i => i.toLowerCase());
    const recipeLower = (recipe.name + (recipe.instructions || '')).toLowerCase();
    
    switch (restriction) {
      case 'Vegan':
        return !ingredientsLower.some(ing =>
          ing.includes('meat') || ing.includes('chicken') ||
          ing.includes('fish') || ing.includes('egg') ||
          ing.includes('milk') || ing.includes('cheese') ||
          ing.includes('butter') || ing.includes('cream')
        );
      case 'Vegetarian':
        return !ingredientsLower.some(ing =>
          ing.includes('meat') || ing.includes('chicken') ||
          ing.includes('fish') || ing.includes('beef')
        );
      case 'Gluten-Free':
        return !ingredientsLower.some(ing =>
          ing.includes('wheat') || ing.includes('flour') ||
          ing.includes('bread') || ing.includes('pasta')
        );
      case 'Dairy-Free':
        return !ingredientsLower.some(ing =>
          ing.includes('milk') || ing.includes('cheese') ||
          ing.includes('butter') || ing.includes('cream')
        );
      case 'Keto':
        return recipeLower.includes('keto') || !ingredientsLower.some(ing =>
          ing.includes('sugar') || ing.includes('rice') ||
          ing.includes('pasta') || ing.includes('bread')
        );
      case 'Paleo':
        return !ingredientsLower.some(ing =>
          ing.includes('dairy') || ing.includes('grain') ||
          ing.includes('legume') || ing.includes('processed')
        );
      case 'Low-Carb':
        return !ingredientsLower.some(ing =>
          ing.includes('rice') || ing.includes('pasta') ||
          ing.includes('bread') || ing.includes('potato')
        );
      case 'High-Protein':
        return ingredientsLower.some(ing =>
          ing.includes('chicken') || ing.includes('beef') ||
          ing.includes('fish') || ing.includes('egg') ||
          ing.includes('protein')
        );
      default:
        return true;
    }
  }
}

export default RecipeService;
